File: 贪心算法/763.划分字母区间.java
class Solution {

    public List<Integer> partitionLabels(String S) {
        char[] s=S.toCharArray();
        int n=s.length;
        //先记住每个字母最后出现的位置
        int[] last = new int[26];
        for(int i=0;i<n;i++){
            last[s[i]-'a']=i;
        }
        List<Integer> res = new ArrayList<>();
        //类似跳跃游戏II的思路
        //确定当前区间的右边界,如果到最右边界了，就划分一个区间，并更新左边界
        int start=0,end=0;
        for(int i=0;i<n;i++){
            end=Math.max(end,last[s[i]-'a']);
            if(i==end){
                res.add(end-start+1);
                start=i+1;
            }
        }
        return res;
    }
}

File: 贪心算法/763.划分字母区间.go
func partitionLabels(s string) []int {
    last:=make([]int,26)
    for i:=0;i<len(s);i++{
        last[s[i]-'a']=i
    }
    res:=[]int{}
    start,end:=0,0
    for i:=0;i<len(s);i++{
        end=max(end,last[s[i]-'a'])
        if i==end{
            res=append(res,end-start+1)
            start=i+1
        }
    }
    return res
}